Build fails with LTO

I tried to build with the following *FLAGS to optimize the build: `-flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing`

Note the -Werror=* flags are used to help detect cases where the compiler tries to optimize by assuming UB cannot exist in the source code -- if it does exist, ordinarily the code would be miscompiled, and this says to make the miscompilation a fatal error.

I got this error:
```
FAILED: _core.cpython-311-x86_64-linux-gnu.so 
: && /usr/bin/x86_64-pc-linux-gnu-g++ -fPIC -march=native -fstack-protector-all -O2 -pipe -fdiagnostics-color=always -frecord-gcc-switches -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-clash-protection -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing  -Wformat -Werror=format-security -O3 -DNDEBUG  -Wl,-O1 -Wl,--as-needed -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing -Wl,--defsym=__gentoo_check_ldflags__=0   -flto -shared  -o _core.cpython-311-x86_64-linux-gnu.so CMakeFiles/_core.dir/src/application.cpp.o CMakeFiles/_core.dir/src/contours.cpp.o CMakeFiles/_core.dir/src/equal.cpp.o CMakeFiles/_core.dir/src/fcn.cpp.o CMakeFiles/_core.dir/src/functionminimum.cpp.o CMakeFiles/_core.dir/src/functionminimum_extra.cpp.o CMakeFiles/_core.dir/src/functionminimum_pickle.cpp.o CMakeFiles/_core.dir/src/hesse.cpp.o CMakeFiles/_core.dir/src/lasymmatrix.cpp.o CMakeFiles/_core.dir/src/lavector.cpp.o CMakeFiles/_core.dir/src/machineprecision.cpp.o CMakeFiles/_core.dir/src/main.cpp.o CMakeFiles/_core.dir/src/migrad.cpp.o CMakeFiles/_core.dir/src/minimumstate.cpp.o CMakeFiles/_core.dir/src/minos.cpp.o CMakeFiles/_core.dir/src/minuitparameter.cpp.o CMakeFiles/_core.dir/src/print.cpp.o CMakeFiles/_core.dir/src/printimpl.cpp.o CMakeFiles/_core.dir/src/scan.cpp.o CMakeFiles/_core.dir/src/simplex.cpp.o CMakeFiles/_core.dir/src/strategy.cpp.o CMakeFiles/_core.dir/src/usercovariance.cpp.o CMakeFiles/_core.dir/src/userparameterstate.cpp.o CMakeFiles/_core.dir/src/usertransformation.cpp.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/AnalyticalGradientCalculator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/BFGSErrorUpdator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/DavidonErrorUpdator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/HessianGradientCalculator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/ExternalInternalGradientCalculator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/InitialGradientCalculator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaEigenValues.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaInnerProduct.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaInverse.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaOuterProduct.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaSumOfElements.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/LaVtMVSimilarity.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MPIProcess.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MinimumBuilder.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnApplication.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnContours.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnCovarianceSqueeze.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnFcn.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnFunctionCross.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnGlobalCorrelationCoeff.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnHesse.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnLineSearch.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnMachinePrecision.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnMinos.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnParabolaFactory.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnParameterScan.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnPlot.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnPosDef.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnPrint.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnSeedGenerator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnStrategy.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnScan.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnTiny.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnTraceObject.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnUserFcn.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnUserParameterState.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnUserParameters.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/MnUserTransformation.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/ModularFunctionMinimizer.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/NegativeG2LineSearch.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/Numerical2PGradientCalculator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SimplexSeedGenerator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SimplexBuilder.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SimplexParameters.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SinParameterTransformation.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/ScanBuilder.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SqrtLowParameterTransformation.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/SqrtUpParameterTransformation.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/VariableMetricBuilder.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/VariableMetricEDMEstimator.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnbins.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mndasum.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mndaxpy.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnddot.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mndscal.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mndspmv.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mndspr.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnlsame.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnteigen.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mntplot.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnvert.cxx.o CMakeFiles/_core.dir/extern/root/math/minuit2/src/mnxerbla.cxx.o   && cd /var/tmp/portage/dev-python/iminuit-2.25.2/work/iminuit-2.25.2/build/cp311-cp311-manylinux_2_38_x86_64 && /usr/bin/x86_64-pc-linux-gnu-strip /var/tmp/portage/dev-python/iminuit-2.25.2/work/iminuit-2.25.2/build/cp311-cp311-manylinux_2_38_x86_64/_core.cpython-311-x86_64-linux-gnu.so
/var/tmp/portage/dev-python/iminuit-2.25.2/work/iminuit-2.25.2/src/type_caster.hpp:10:8: error: type ‘struct type_caster’ violates the C++ One Definition Rule [-Werror=odr]
   10 | struct type_caster<std::vector<Value>> {
      |        ^
/usr/lib/python3.11/site-packages/pybind11/include/pybind11/stl.h:215:8: note: a type with different bases is defined in another translation unit
  215 | struct type_caster<std::vector<Type, Alloc>> : list_caster<std::vector<Type, Alloc>, Type> {};
      |        ^
lto1: some warnings being treated as errors
lto-wrapper: fatal error: /usr/bin/x86_64-pc-linux-gnu-g++ returned 1 exit status
compilation terminated.
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Downstream bug report: https://bugs.gentoo.org/856778
Full build log: [build.log](https://github.com/scikit-hep/iminuit/files/14552638/build.log)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -87,8 +87,16 @@ if(MSVC)
   target_compile_options(_core PRIVATE /std:c++14 /Y-)
 else()
   # lots of warnings from clang and gcc
-  target_compile_options(_core PRIVATE -Wall -Wextra -pedantic -Werror
-                                       -fstrict-aliasing)
+  target_compile_options(
+    _core
+    PRIVATE -Wall
+            -Wextra
+            -pedantic
+            -Werror
+            -fstrict-aliasing
+            -Werror=odr
+            -Werror=selector-type-mismatch
+            -Werror=strict-aliasing)
 endif()
 target_include_directories(_core PRIVATE extern/root/math/minuit2/inc)
 set_target_properties(_core PROPERTIES VISIBILITY_INLINES_HIDDEN ON)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -85,8 +85,8 @@ set(SOURCES_B
 pybind11_add_module(_core MODULE ${SOURCES_A} ${SOURCES_B})
 if(MSVC)
   target_compile_options(_core PRIVATE /std:c++14 /Y-)
-else()
-  # lots of warnings from clang and gcc
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  # lots of warnings from gcc, including odr violations
   target_compile_options(
     _core
     PRIVATE -Wall
@@ -95,8 +95,12 @@ else()
             -Werror
             -fstrict-aliasing
             -Werror=odr
-            -Werror=selector-type-mismatch
+            -Werror=lto-type-mismatch
             -Werror=strict-aliasing)
+else()
+  # lots of warnings from clang
+  target_compile_options(_core PRIVATE -Wall -Wextra -pedantic -Werror
+                                       -fstrict-aliasing)
 endif()
 target_include_directories(_core PRIVATE extern/root/math/minuit2/inc)
 set_target_properties(_core PROPERTIES VISIBILITY_INLINES_HIDDEN ON)

--- a/src/type_caster.hpp
+++ b/src/type_caster.hpp
@@ -1,4 +1,3 @@
-#include <algorithm>
 #include <pybind11/numpy.h>
 #include <pybind11/pybind11.h>
 #include <vector>
@@ -6,9 +5,9 @@
 namespace pybind11 {
 namespace detail {
 
-template <typename Value>
-struct type_caster<std::vector<Value>> {
-  using vec_t = std::vector<Value>;
+template <typename Value, typename Alloc>
+struct type_caster<std::vector<Value, Alloc>> {
+  using vec_t = std::vector<Value, Alloc>;
   using value_conv = make_caster<Value>;
   using size_conv = make_caster<std::size_t>;
 

